channels/rdpgfx/server/rdpgfx_main.c

```diff
@@ -1170,31 +1170,30 @@ static UINT rdpgfx_recv_caps_advertise_pdu(RdpgfxServerContext* context,
 
 	Stream_Read_UINT16(s, pdu.capsSetCount); /* capsSetCount (2 bytes) */
 
-	if (pdu.capsSetCount > 3)
-	{
-		/* According to the latest spec, capsSetCount <= 3 */
-		WLog_ERR(TAG, "capsSetCount is greater than 3: %u", pdu.capsSetCount);
-		return ERROR_INVALID_DATA;
-	}
-
-	if (Stream_GetRemainingLength(s) < (pdu.capsSetCount * 12))
+	if (Stream_GetRemainingLength(s) < (pdu.capsSetCount * RDPGFX_CAPSET_SIZE))
 	{
 		WLog_ERR(TAG, "not enough data!");
 		return ERROR_INVALID_DATA;
 	}
 
+	capsSets = calloc(pdu.capsSetCount, RDPGFX_CAPSET_SIZE);
+
+	if (!capsSets)
+		return ERROR_OUTOFMEMORY;
+
 	pdu.capsSets = (RDPGFX_CAPSET*) capsSets;
 
 	for (index = 0; index < pdu.capsSetCount; index++)
 	{
-		capsSet = &(pdu.capsSets[index]);
+		RDPGFX_CAPSET* capsSet = &(pdu.capsSets[index]);
```

Comment:
I notice the declare of capsSet is moved into this code block. Some compiler might ask everything to be declared at very beginning of a function for .c files. If we don't want to be compatible with such compiler, then it's just OK.